### PR TITLE
Support prefix object in list result

### PR DESCRIPTION
**Pull request by medb**

### CODE REVIEW

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java [hunk 1]
```diff
@@ -1494,19 +1500,44 @@ private static long getMaxRemainingResults(
     List<String> listedPrefixes = new ArrayList<>();
     String nextPageToken =
         listStorageObjectsAndPrefixesPage(listObject, listOptions, listedObjects, listedPrefixes);
+    List<GoogleCloudStorageItemInfo> objectInfos =
+        getGoogleCloudStorageItemInfos(
+            bucketName, objectNamePrefix, listOptions, listedPrefixes, listedObjects);
+    return new ListPage<>(objectInfos, nextPageToken);
+  }
+
+  private List<GoogleCloudStorageItemInfo> getGoogleCloudStorageItemInfos(
```
Comment (hongyegong) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java:1509]: Can this method be reused in the listObjectNames method above? I find core logic are pretty similar.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/ListObjectOptions.java [hunk 2]
```diff
@@ -42,6 +43,9 @@ public static Builder builder() {
   @Nullable
   public abstract String getDelimiter();
 
+  /** Whether to include prefix object in the result. */
+  public abstract boolean isIncludePrefix();
```
Comment (hongyegong) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/ListObjectOptions.java:47]: would isIncludePrefixEnabled be a better name? isIncludeNoun sounds not right.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java (hunk 1 above)
Comment (medb) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java:1509]: While logic is similar, `listObjectNames` does not create `GoogleCloudStorageItemInfo` objects to improve performance so it's hard to reuse the same code between these 2 methods.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/ListObjectOptions.java (hunk 2 above)
Comment (medb) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/ListObjectOptions.java:47]: Yeah, I don't like `Enabled` suffixes that we have in some places, I guess better name will be `shouldIncludePrefix` but it's not supporter by AutoValue. I can try to drop `is` prefix from getter, if it works?

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java [hunk 3]
```diff
@@ -1412,13 +1417,27 @@ private static long getMaxRemainingResults(
         listedObjects,
         listedPrefixes);
 
+    // Size to accommodate listed prefixes, objects and prefix object
+    List<String> objectNames = new ArrayList<>(listedPrefixes.size() + listedObjects.size() + 1);
+
+    // Add prefix object name if necessary
+    if (listOptions.isIncludePrefix()
```
[the remark this answers is not in the document]
Comment (medb) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java:1424]: Added comments.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/ListObjectOptions.java (hunk 2 above)
Comment (hongyegong) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/ListObjectOptions.java:47]: sounds good.

**Review by hongyegong — APPROVED** [review]

**Pull request merged**